klippy/extras/palette2.py

```diff
@@ -0,0 +1,610 @@
+# Palette 2 MMU support, Firmware 9.0.9 and newer supported only!
+#
+# Copyright (C) 2021 Clifford Roche <clifford.roche@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+import logging
+import os
+import threading
+import time
+import serial
+
+from serial import SerialException
+
+try:
+    from queue import Queue, Empty
+except ImportError:
+    from Queue import Queue, Empty
+
+COMMAND_HEARTBEAT = "O99"
+COMMAND_CUT = "O10 D5"
+COMMAND_CLEAR = [
+    "O10 D5",
+    "O10 D0 D0 D0 DFFE1",
+    "O10 D1 D0 D0 DFFE1",
+    "O10 D2 D0 D0 DFFE1",
+    "O10 D3 D0 D0 DFFE1",
+    "O10 D4 D0 D0 D0069"]
+COMMAND_FILENAME = "O51"
+COMMAND_FILENAMES_DONE = "O52"
+COMMAND_FIRMWARE = "O50"
+COMMAND_PING = "O31"
+COMMAND_SMART_LOAD_STOP = "O102 D1"
+
+HEARTBEAT_SEND = 5
+HEARTBEAT_TIMEOUT = 11
+SETUP_TIMEOUT = 300
+
+SERIAL_TIMER = 0.1
+AUTOLOAD_TIMER = 5.
+
+INFO_NOT_CONNECTED = "Palette 2 is not connected, connect first"
+
+
+class Palette2:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.gcode_move = self.printer.load_object(config, 'gcode_move')
+        self.gcode = self.printer.lookup_object("gcode")
+        self.printer.register_event_handler("klippy:ready", self.handle_ready)
+        self.gcode.register_command(
+            "PALETTE_CONNECT", self.cmd_Connect, desc=self.cmd_Connect_Help)
+        self.gcode.register_command(
+            "PALETTE_DISCONNECT",
+            self.cmd_Disconnect,
+            desc=self.cmd_Disconnect_Help)
+        self.gcode.register_command(
+            "PALETTE_CLEAR", self.cmd_Clear, desc=self.cmd_Clear_Help)
+        self.gcode.register_command(
+            "PALETTE_CUT", self.cmd_Cut, desc=self.cmd_Cut_Help)
+        self.gcode.register_command(
+            "PALETTE_SMART_LOAD",
+            self.cmd_Smart_Load,
+            desc=self.cmd_Smart_Load_Help)
+        self.serial = None
+        self.serial_port = config.get("serial")
+        if not self.serial_port:
+            raise config.error("Invalid serial port specific for Palette 2")
+        self.baud = config.getint("baud", default=250000)
+        self.feedrate_splice = config.getfloat(
+            "feedrate_splice", 0.8, minval=0., maxval=1.)
+        self.feedrate_normal = config.getfloat(
+            "feedrate_normal", 1.0, minval=0., maxval=1.)
+        self.auto_load_speed = config.getint("auto_load_speed", 2)
+
+        # Omega code matchers
+        self.omega_header = [None] * 9
+        omega_handlers = ["O" + str(i) for i in range(33)]
+        for cmd in omega_handlers:
+            func = getattr(self, 'cmd_' + cmd, None)
+            desc = getattr(self, 'cmd_' + cmd + '_help', None)
+            if func:
+                self.gcode.register_command(cmd, func, desc=desc)
+            else:
+                self.gcode.register_command(cmd, self.cmd_OmegaDefault)
```

Comment:
What do all these commands do?  Is there a reason they are registered with the main gcode command handler?

Also, for something like this, it is a good idea to update docs/G-Codes.md .

Reply:
The Palette 2 works by embedded a services of device code into the gcode file, and reading them in the GCode stream. These are referred to as "O" or "Omega" codes.

As an example, this is what the intro sequence to my test print would look like (there are codes throughout the file though).

```
O21 D0014
O22 D373fd85d66ac76f6
O23 D0001
O24 D0000
O25 D0 D11565c0Dark_Blue_Filament D1ffd700Yellow_Filament D0
O26 D0005
O27 D0007
O28 D0001
O29 D0000
O30 D1 D42cc5f08
O30 D2 D438b30c2
O30 D1 D450a70f6
O30 D2 D45518aab
O30 D1 D458af459
O32 D11 D0000 D0000 D0000
O1 DMosaic-Keychain D0000115f
M0
; Sliced by CANVAS
;
; *** Actual Slicing Settings As Used ***
;

...

```

All OCodes read from the file, need to be passed to the device. Some OCodes require additional processing.

As for documentation though, as you can see, these codes are really not meant to be input on a terminal. They are highly statefull, calling in the wrong order, with incomplete data, etc, they would be incredibly difficult to document and of no practical value.

Instead special commands like PALETTE_CUT, CLEAR are exposed as separate commands to navigate this. I did, however, not add them to documentation yet.

Reply:
I added documentation for the user custom GCodes.